### DISCUSSION TITLE
Elsa 2.5.0 🍦

### DISCUSSION THREAD

**Discussion by sfmskywalker** [elsa-workflows/elsa-core#2741] [Design & tech]
### What's Changed

**Trigger Indexing**

A major change was introduced that changes the way how workflow triggers are indexed.

Originally, ALL workflow blueprints would be loaded into memory during application startup, after which the triggers were being indexed.
This however poses a serious startup performance issue when the system contains more than a few hundred workflow definitions. Some users reported a startup time of around 15 seconds when testing with a set of around 1500 workflow definitions.

Elsa needs to be able to handle **any** number of workflow definitions, so loading these into memory is not a good idea.
To fix this, a change was necessary.

Instead of indexing triggers from workflow definitions (typically stored in the database) during startup, you can now configure Elsa to skip this step and instead rely on indexing of triggers as workflow definitions get published:

```csharp
services
    .AddElsa(elsa => elsa   
        // Optionally opt-out of indexing workflows stored in the database.
        // These will be indexed when published/unpublished/deleted, so no need to do it during startup.
        // Unless you have existing workflow definitions in the DB for which no triggers have yet been created.
        .ExcludeWorkflowProviderFromStartupIndexing<DatabaseWorkflowProvider>()
    )
```

In addition, workflow triggers are no longer stored in memory. Instead, they are stored in the database.
If you are using the EF Core persistence provider and are maintaining your own migrations, then please take a look at the changed schema (see new migrations in the repo).

### What's New

** UserTask **

A new service called `IUserTaskService` was added to the user task module that simplifies querying the system for available user actions for a given workflow instance or all available user tasks.

This is helpful when you e.g. need to render a list of workflow instances with available actions.

### What's Improved

- Designer updated to use Tailwind CSS 3
- Workflow Registry HTTP endpoint no longer returns child activities of composite activities, improving performance in the designer (e1d42e80c80a14c55eb584b74ecb0409c0605a9c)
- Persist workflow after status change and before executing. This prevents an issue where the app or container might crash and the workflow is still in e.g. the Suspended state, which are never resumed automatically (7c1da5b3d92264c6e98e73491743dc07ed4a56be)
- Adding convenience Extension Methods for setting properties on Rebus Activities, following the 'with*' naming convention (#2712) (Adding convenience Extension Methods for setting properties on Rebus Activities, following the 'with*' naming convention (#2712))

### What's Fixed

- Fix overriding parents of sub composites (ba8fb6837340cc5009c1ba5ebe75bd5c9cc687a4)
- Update Elsa.Client workflow definition models to match Elsa.Server (#2716) (efd9f0d446f1c56b45aa84bad30d496936d051dc)
- Fixed journal time display. Fixed minor visual issues. (#2722) (180f36bb78c7411a55029bcf964f040fae1a21bf)

<hr /><em>This discussion was created from the release <a href='https://github.com/elsa-workflows/elsa-core/releases/tag/2.5.0'>Elsa 2.5.0 🍦</a>.</em>

**Comment by Nico-VanHaaster** [comment]
@sfmskywalker I am just curious on the ramifactions of using the `ExcludeWorkflowProviderFromStartupIndexing<DatabaseWorkflowProvider>();`

Currently we have a multi-tenant hosted workflow solution with quite a few Cron based triggers. It has been noted that the workflow system can take up to 10 minutes to startup. I have worked this out to your comments above about how the `ExcludeWorkflowProviderFromStartupIndexing<DatabaseWorkflowProvider>();` can be used to improve startup times.

For some background. All our workflows are database driven. We don't use any Code workflows and they are all sourced from the database directly (EF Storage). From my understanding as we are using the UIs \ APIs for creating \ publishing \ unpublishing etc that disabling this provider from indexing should be fine.

Can you think of any other risks with this approach?

Cheers,
Nico

**Reply by sfmskywalker** [reply]
No, should be fine. Just keep in mind that triggers are only created upon publishing workflows instead during application startup. If you always use the designer and or REST APIs to manage workflow definitions, then you’ll definitely want to enable this “skip database workflow provider based trigger indexing”, especially if you have big and or many workflows 👍🏻

**Reply by Nico-VanHaaster** [reply]
Great thank you for the quick response. I have disabled this and now the system is fully started within 30seconds which is a fair bit faster than before.

**Reply by sfmskywalker** [reply]
30 seconds is still slow, but I have seen this too. In my case, it’s because of Rebus setting up workers and what not. Also the recreation of Quartz.NET timer triggers can take a long time because this happens one by one. Which is not actually necessary to do when using Quartz.NET persistence. I fixed this in V3.

**Reply by Nico-VanHaaster** [reply]
Yeah 30 seconds is manageable as this is clustered in a Kubernetes across multiple regions. We also have some of our code that gets in the way of startup time. We are also using RebusCacheSignalars and Distributed Caching using Azure Blob provider plus Clustering. It may be setup a bit incorrectly.
